### PR TITLE
ports/renesas-ra/boards/VK-RA6M5: Last New Board.

### PR DESCRIPTION
**Pull request by mbedNoobNinja**
In the context of ongoing discussions (#10595) & (#10752) and the mentioned intention for a new ports, this is the **3-rd** uPy ready board candidat.

`ports/renesas-ra/boards/`**`VK-RA6M5`:** Adding another new dev board
`ports/renesas-ra/boards:` Based on new **RA6M5** processor
`ports/renesas-ra:` Part of existing **RA** family.

[Tests.log](https://github.com/micropython/micropython/files/10899348/M5_Tests.log)

### CODE REVIEW

File: ports/renesas-ra/Makefile
Comment (iabdalkader) [inline ports/renesas-ra/Makefile:59]: Can you please filter this on the series ? So other RA6M5 boards can build ?

```suggestion
ifeq ($(BOARD),RA6M5)
EXPECTED_FSP_HASH = 9713f5b5c3d8bea5f0242feb30c1a57c89ae506a lib/fsp (v2.4.0)
```

File: ports/renesas-ra/Makefile
Comment (iabdalkader) [inline ports/renesas-ra/Makefile:54]: I think all of those checks in the Makefile (like `ifeq  ($(BOARD), some_EK)`) should all be removed to simplify the Makefile, all the variables are already there:
```
BOARD_LOW  = (shell echo $(BOARD) | tr '[:upper:]' '[:lower:]')
CMSIS_MCU_CAP = $(CMSIS_MCU)
CMSIS_MCU_LOWER = $(shell echo $(CMSIS_MCU) | tr '[:upper:]' '[:lower:]')
USE_FSP_LPM ?=0 (or 1) // This should be moved to `mpconfigboard.mk` of each board.
```
Note there's a typo in the clicker board:
```
ifeq ($(BOARD),RA4M1_CLICKER)
BOARD_LOW = ra4m1_ek <- Should be ra4m1_clicker
```

File: ports/renesas-ra/Makefile
Comment (iabdalkader) [inline ports/renesas-ra/Makefile:352]: Can you please filter on the series and in `machine_dac.c` use something like:
```
#if MICROPY_HW_ENABLE_DAC
machine_dac_init()...
#endif
```

This way other `RA6M5` boards can use this without editing the Makefile.

File: ports/renesas-ra/Makefile
Comment (iabdalkader) [inline ports/renesas-ra/Makefile:122]: Can you use the existing naming convention for the series ? i.e `RA4W1_hal.h`, `RA6M3_hal.h` and `RA6M5_hal.h`. It seems the only difference between the existing config files and the new ones you're adding is the default SCI debug channel, which can be defined in `mpconfigboard.h`. Also note that  `DEBUG_TX_PIN` and `DEBUG_RX_PIN` are never used anywhere so any defaults there (the EK maybe ?) should be fine.

File: ports/renesas-ra/Makefile
Comment (iabdalkader) [inline ports/renesas-ra/Makefile:372]: I suggest you use a `$(wildcard $(BOARD_DIR)/*.c)` here so other boards can have `board_init.c` files.

File: ports/renesas-ra/Makefile
Comment (iabdalkader) [inline ports/renesas-ra/Makefile:483]:
```suggestion
ifeq ($(CMSIS_MCU),RA6M5)
```

File: ports/renesas-ra/Makefile
Comment (mbedNoobNinja) [reply ports/renesas-ra/Makefile:122]: Yes I tried to use the default one hals (`*_hal.h`) but the default pins of the EK boards are other and even I select the channel in the `mpconfigbord.h` it is still pointing to the wrong ones. If you have a beter mechanism for same channel to select other alternative pins I am listening ...

File: ports/renesas-ra/Makefile
Comment (mbedNoobNinja) [reply ports/renesas-ra/Makefile:372]: Every other bords **will have** `board_init.c` and `board_led.c`, except the VK ones. These files exist only if a given board has a BSP in the `FSP`, so every board not part of some kind of program or arangment with Renesas will be wiped out  with next upgrade of the FSP. It si pointless to add them to all boards unless you are sure that all boards will be included in every next new FSP release.

File: ports/renesas-ra/Makefile
Comment (mbedNoobNinja) [reply ports/renesas-ra/Makefile:483]: OK that makes sense, It will be fixed.

File: ports/renesas-ra/Makefile
Comment (mbedNoobNinja) [reply ports/renesas-ra/Makefile:434]: I haven't samples of other boards to test with, so I am not sure if it will work on them If it is released for the series. This only @TakeoTakahashi2020 can confirm. Ask him.

File: ports/renesas-ra/Makefile
Comment (iabdalkader) [reply ports/renesas-ra/Makefile:122]: > If you have a beter mechanism for same channel to select other alternative pins I am listening ...

Well like I said the `DEBUG_RX_PIN` and `DEBUG_TX_PIN` pins are not really used anyway (grep and you won't find those defines used anywhere) so I'm not sure why you want to override them, however if you need to override them for some reason it's possible, but for now I think we just need to override the default debug channel.

There's more than one way to do override `DEBUG_CH` but I think there's a file that exists solely for this purpose, which is `ra*_ek_conf.h` and for some reason it's not included anywhere, but we can use it. If you just rename `vk_ra6m1_conf.h` to `ra6m5_ek_conf.h` and include that file in `RA6M5_hal.h` then you have the debug channel defined (and of course every `RAx_hal.h` should include its `rax_ek_conf.h`). You can also use the same file to define `USE_DBG_PRINT` when needed.

File: ports/renesas-ra/Makefile
Comment (mbedNoobNinja) [reply ports/renesas-ra/Makefile:59]: Yes that makes sense, It will be fixed.

File: ports/renesas-ra/Makefile
Comment (iabdalkader) [reply ports/renesas-ra/Makefile:372]: Sorry didn't notice the `ifneq`, I think the exception should be the official EK boards (because every new board will Not have those files, so will have to be added to the Makefile) ,so I think the condition should check for official boards not custom boards, or better maybe just  ?

```
EK_BOARD_DIR = $(HAL_DIR)/ra/board/$(BOARD_LOW)
ifneq ($(wildcard $(TOP)/$(EK_BOARD_DIR)/.),)
HAL_SRC_C += $(addprefix $(EK_BOARD_DIR)/,\
	board_init.c \
	board_leds.c \
	)
endif
```

File: ports/renesas-ra/Makefile
Comment (mbedNoobNinja) [reply ports/renesas-ra/Makefile:122]: Yes I do it with a separate file because i was afraid it will break the compile of the existing boards and I hadn't noticed it is not used. OK if you insist to be in 1 file it will be, I will fix it.

File: ports/renesas-ra/Makefile
Comment (mbedNoobNinja) [reply ports/renesas-ra/Makefile:372]: Well I have power over VK boards, for the EK ask @TakeoTakahashi2020.

File: ports/renesas-ra/Makefile
Comment (mbedNoobNinja) [reply ports/renesas-ra/Makefile:352]: Again DAC is not confirmed it works on all boards, so ot will remain ON for those it is tested with.

File: ports/renesas-ra/Makefile
Comment (iabdalkader) [reply ports/renesas-ra/Makefile:122]: > OK if you insist to be in 1 file it will be, I will fix it.

Not insisting but IMHO I think it's much cleaner this way, specific board names should Not be in the Makefile I know there were some board names in the Makefile before, but this doesn't mean we can't clean it up. My goal here is to help you make these changes as generic as possible, this will hopefully speed up the merging of this PR and allow others to reuse your work here.

File: ports/renesas-ra/Makefile
Comment (iabdalkader) [reply ports/renesas-ra/Makefile:352]: It will not be on/initialized unless `MICROPY_HW_DAC0` or `MICROPY_HW_DAC1` is defined in `mpconfigboard.h` (Note you disabled by default in `mpconfigboard_common.h`), building the drivers in `ra/` should not affect boards that don't use it.

File: ports/renesas-ra/Makefile
Comment (mbedNoobNinja) [reply ports/renesas-ra/Makefile:54]: Well the design of the Makefile is what it is, I am just following it to integrate the VK boards. I don't know if I have the right to alter it. I think Mr. @TakeoTakahashi2020 has the last word here. Ask him why that is.

File: ports/renesas-ra/Makefile
Comment (iabdalkader) [reply ports/renesas-ra/Makefile:483]: On second thought, since the boards are tightly coupled to the FSP version used in the e2studio, I have a better suggestion here, maybe just set `EXPECTED_FSP_HASH` in `mpconfigboard.mk` ? This way every board can override the default *if needed*. In the Makefile you can just set the default if it's not defined:

```Makefile
EXPECTED_FSP_HASH ?= $(patsubst +%,%,$(shell git -C ../.. submodule status --cached lib/fsp))
```

File: ports/renesas-ra/Makefile
Comment (iabdalkader) [reply ports/renesas-ra/Makefile:122]: Oh no please take your time, I'm currently testing your PR with these suggested changes, and reporting back anything that I think might be helpful ;)

File: ports/renesas-ra/Makefile
Comment (mbedNoobNinja) [reply ports/renesas-ra/Makefile:483]: Great Idea I will be happy to fix it that way.

File: ports/renesas-ra/Makefile
Comment (TakeoTakahashi2020) [reply ports/renesas-ra/Makefile:54]: Great idea, I am welcome to simplify Makefile. 
Regarding RA4M1_CLICKER's typo,  because it is not official EK board and not included in FSP, EK board's file is just referred instead.  However it's same situation as VK board and it should be solved by same way of VK boards.  I can check EK board and Clicker together, so please change as you propose.

File: ports/renesas-ra/Makefile
Comment (TakeoTakahashi2020) [reply ports/renesas-ra/Makefile:434]: Please change it.  I can check it on EK_RA4W1 and report if there are any issues.

File: ports/renesas-ra/Makefile
Comment (iabdalkader) [reply ports/renesas-ra/Makefile:54]: @TakeoTakahashi2020 I see the issue with `RA4M1_CLICKER` board, thanks for the insight. Even though `RA4M1_CLICKER` seems to build without linking `board_init.c` and `board_leds.c`, but those sources can still be added like so:

```Makefile
BOARD_LOW ?= $(shell echo $(BOARD) | tr '[:upper:]' '[:lower:]')
```

Then in `boards/RA4M1_CLICKER/mpconfigboard.mk` you can just set

```Makefile
BOARD_LOW=ra4m1_ek
```

I'd suggest a more descriptive name, like `FSP_BOARD_NAME`

Comment (TakeoTakahashi2020) [reply ports/renesas-ra/Makefile:54]: @iabdalkader , Thanks for your great review and suggetion.  I did build check quickly.  Yes, you are right.  The build is successful without board_init.c and board_leds.c.  I also think FSP_BOARD_NAME makes sense.
I completely agree with your idea.  I am happy to be able to simplify. Thanks.

File: ports/renesas-ra/Makefile
Comment (iabdalkader) [reply ports/renesas-ra/Makefile:483]: Just noticed that switching between two boards that require two different FSPs *with* parallel builds (`make -j<n>`) seems to fail on the first run (only the first run), so it seems that some objects start building before FSP is switched. Making the FSP check an order dependency of `$(OBJ)` seems to fix it:

```Makefile
$(OBJ): | CHECK_FSP_VER $(GEN_PINS_HDR)
```

File: ports/renesas-ra/Makefile
Comment (TakeoTakahashi2020) [reply ports/renesas-ra/Makefile:372]: @mbedNoobNinja , @iabdalkader ,  I have confirmed that the building of RA4M1_CLICKER is successful with EK_BOARD_DIR and BOARD_LOW before the code edited by @iabdalkader.  So FSP_BOARD_DIR and FSP_BOARD_NAME also shoud be OK.  I agree with this change.

File: ports/renesas-ra/Makefile
Comment (TakeoTakahashi2020) [reply ports/renesas-ra/Makefile:483]: What a great idea!  It was a headache for me. Thank you @iabdalkader and @mbedNoobNinja .

**Comment by mbedNoobNinja** [comment]
Hi @iabdalkader and @TakeoTakahashi2020 I finally found time to make the changes you suggested, but I am currently unaware why I am getting check commit message formatting. Any idea?

**Comment by iabdalkader** [comment]
> why I am getting check commit message formatting. Any idea?

No idea really, it seems to be complaining about commits not in this PR, you could try rebasing on master.

**Comment by mbedNoobNinja** [comment]
You are right, rebase solves the issue!

**Review by iabdalkader — APPROVED** [review]

**Comment by iabdalkader** [comment]
@mbedNoobNinja This looks really good thanks! I just have a couple of minor comments

* You may want remove all of your `e2studio` project files (`.cproject`, `.project`, `configuration.xml`, `makefile.defs` etc..) as they're not used in micropython and none of the other boards include them.

* You may want to double check your board's RTC clock source in `mpconfigboard.h`, seems in `hal_data.c` you set it to `RTC_CLOCK_SOURCE_LOCO` 🤷🏼‍♂️ 

* None of the `boards/<BOARD>/xxx_conf.h` files is included, so `DEBUG_CH` is not set, I'd suggest to just remove all those files and the debug channel can be defined in `mpconfigboard.mk` with something like `CFLAGS+=-DDEBUG_CH=0` if needed.

Now a question for you and @TakeoTakahashi2020 has this been tested on any actual board ? Did you get a working REPL on the UART ? I created a new project, selected the device, configured FSP, set the heap and stack, added LPM, RTC, SCI UART and ICU  stacks (set the `user_uart_callback`, and enabled all 4 IRQs), fixed the linker script, and I get as far as executing friendly repl, but I don't see any repl output on the UART, I wonder if I'm missing something ? What exactly needs to be configured in `e2studio` to generate a project that works with this port ?

**Comment by TakeoTakahashi2020** [comment]
@iabdalkader  I could build for all EK and CLICKER board without any errors and got the same result of test on the actual board as previous version.  Thanks @mbedNoobNinja .

Regarding removing `boards/<BOARD>/xxx_conf.h`, I agree with your suggestion.

Regarding UART for REPL, are `MICROPY_HW_UART_REPL' and `MICROPY_HW_UART_REPL_BAUD` defined in `board/<BOARD>/mpconfigboard.h?

**Comment by iabdalkader** [comment]
> Regarding UART for REPL, are `MICROPY_HW_UART_REPL` and `MICROPY_HW_UART_REPL_BAUD` defined in `board/<BOARD>/mpconfigboard.h`?

Yes, I have this in the `mpconfigboard.h`:

```C
#define MICROPY_HW_UART9_TX         (pin_P602)
#define MICROPY_HW_UART9_RX         (pin_P110)

#define MICROPY_HW_UART_REPL        (HW_UART_9)
#define MICROPY_HW_UART_REPL_BAUD   (115200)
```

And all 4 IRQs are enabled, in the `e2studio` project:
```C
            [0] = rtc_alarm_periodic_isr, /* RTC ALARM (Alarm interrupt) */
            [1] = rtc_alarm_periodic_isr, /* RTC PERIOD (Periodic interrupt) */
            [2] = rtc_carry_isr, /* RTC CARRY (Carry interrupt) */
            [3] = r_icu_isr, /* ICU IRQ0 (External pin interrupt 0) */
            [4] = sci_uart_rxi_isr, /* SCI9 RXI (Received data full) */
            [5] = sci_uart_txi_isr, /* SCI9 TXI (Transmit data empty) */
            [6] = sci_uart_tei_isr, /* SCI9 TEI (Transmit end) */
            [7] = sci_uart_eri_isr, /* SCI9 ERI (Receive error) */
```

The callback is set too:
```C
const uart_cfg_t g_uart9_cfg =
{ .channel = 9, .data_bits = UART_DATA_BITS_8, .parity = UART_PARITY_OFF, .stop_bits = UART_STOP_BITS_1, .p_callback = user_uart_callback,
```

**Comment by TakeoTakahashi2020** [comment]
LGTM.  Other settings for UART9 on e2 studio are as followings (however, please note this is for EK-RA6M2):

image

**Comment by iabdalkader** [comment]
It seems the IRQs are not working for some reason (`ra_sci_isr_tx` is never called), if I write directly to the SCI TDR I can see the output. I'm still investigating but it looks like I may be missing something.

**Comment by iabdalkader** [comment]
> So are you OK if I remove that file for all else boards and preserve it only in RA4W1_EK and VK_RA4W1 dirs eventually?

No please keep them and include all of the config files (which is what I initially suggested btw in my comment https://github.com/micropython/micropython/pull/10943#discussion_r1147471564). I only suggested you remove them later,  because you didn't include them in the `RAxxx_hal.h` file (in this PR, I don't know what you're doing/going to do in future PRs :)) so this should be in every `RAxxx_hal.h` file 

```C
#if defined(USE_DBG_PRINT)
#include RA_CFG_H
#endif
```
I hope it's more clear now. That said, I'm happy with this PR as it is, and making any further changes is up to you.

BTW I have TinyUSB support working for this port (tested with RA6M5) it's based on top of this PR, so will send it after this is merged. CC @TakeoTakahashi2020

**Review by iabdalkader — APPROVED** [review]

**Comment by iabdalkader** [comment]
@mbedNoobNinja 

The ADC peripheral base address for RA6M5 is different, it should be:

```C
#if defined(RA6M1) | defined(RA6M2) | defined(RA6M3)
static R_ADC0_Type *adc_reg = (R_ADC0_Type *)0x4005c000;
#elif defined(RA6M5)
static R_ADC0_Type *adc_reg = (R_ADC0_Type *)0x40170000;
#endif
```

**Comment by dpgeorge** [comment]
Thanks for the contribution @mbedNoobNinja, and for the review and discussion @TakeoTakahashi2020 and @iabdalkader .

Since there is a lot of work on this PR to get it ready for merge, I guess we will do this one first and then follow up with the other two (#10595 and #10752) later.

File: ports/renesas-ra/Makefile
Comment (dpgeorge) [inline ports/renesas-ra/Makefile:431]: I think this test will always pass because both these variables are computed by calling `git status`.

Regardless of that, doing git submodule updates like this is very unconventional.  No other port does this so this reneses-ra port should not do it either.  Instead the submodule should just be updated as it normally would be.

In summary, please just remove this section that deals with the hash of the fsp lib.

File: ports/renesas-ra/Makefile
Comment (dpgeorge) [inline ports/renesas-ra/Makefile:508]: Why is RA6M3 support needed if this board is a RA6M5 board?

**Comment by dpgeorge** [comment]
@mbedNoobNinja can you please rebase this on the latest master?  That has renamed some EK boards.

**Comment by mbedNoobNinja** [comment]
Hi @dpgeorge, I will fix the remarks by the end of the day. Reguarding the RA6M3 I missed it on comparison with the M3's PR and really it is not needed here. Currently I am tring to set uPy to use the QSPI as an internal flash FS, instead of the standard MCU's code flash. Does any other board done that, so I can get inspiration from it?

File: ports/renesas-ra/Makefile
Comment (iabdalkader) [reply ports/renesas-ra/Makefile:431]: The `EXPECTED_FSP_HASH` can be defined per-board in `mpconfigboard.mk` files to override the default FSP version (the submodule tag/commit). The problem is the `e2studio` projects generated for every board, depend on the FSP version that was used to generate them. For example support for the new series was introduced in a more recent FSP, but all of the other boards use an older FSP, and our board `fsp-4.0.0`. Ideally FSP should be updated to the latest version and all the projects updated to match it, but that will be a lot of work, or it might be possible to remove the dependency on `e2studio` project files, I did notice that most of the generated files are not really used (only the vector table seem important, and the headers that enable the different stacks).

Comment (dpgeorge) [reply ports/renesas-ra/Makefile:431]: We cannot have different boards requiring a different commit hash of the fsp submodule.

The fsp really needs to be updated to support the new board.  For example this is how stm32lib works when a new ST MCU is supported.

But, does the VK-RA6M5 here need a new fsp?

Comment (iabdalkader) [reply ports/renesas-ra/Makefile:431]: > But, does the VK-RA6M5 here need a new fsp?

It does yes, it seems support for the series was introduced in `2.4.0` and then following releases added important fixes, see the releases below... For Portenta-C33 I based the project on `4.0.0`, I'm okay with regenerating the project, it's still local, but to update the the FSP to any newer version someone has to regenerate all the projects for all of the other boards.

https://github.com/renesas/fsp/releases/tag/v2.4.0

Comment (dpgeorge) [reply ports/renesas-ra/Makefile:431]: > but to update the the FSP to any newer version someone has to regenerate all the projects for all of the other boards.

I don't understand why that's necessary... what files need to be regenerated when the fsp is updated?

Comment (iabdalkader) [reply ports/renesas-ra/Makefile:431]: > I don't understand why that's necessary... what files need to be regenerated when the fsp is updated?

All the files in `boards/*/ra_gen`, `boards/*/ra_cfg`, `boards/*/src`...I'm not sure which files are actually used, to me it seems only the generated vector table and `ra_cfg/*` are important.

Comment (dpgeorge) [reply ports/renesas-ra/Makefile:431]: Can you please explain how they are generated?  Can they be generated as part of the build process?

Comment (TakeoTakahashi2020) [reply ports/renesas-ra/Makefile:431]: Yes, your explanation is correct.  They are generated by e2 studio acoording to the project configuration.  So they can not be generated as part of the MicroPython build process. 
However, basically the FSP will try to keep conpatibility and will keep updating to support new devices.  The latest specific version of FSP might be easy to support renesas-ra port itself.    So if you all could agree to update FSP to the latest one or recent one, I agree with your decision and I am happy to check all EK and Clicker and will fix if there were any issues.

Comment (dpgeorge) [reply ports/renesas-ra/Makefile:431]: OK, in that case, can we try to update the FSP to the latest version?  That task should be done as a separate, independent task with its own PR.

Comment (TakeoTakahashi2020) [reply ports/renesas-ra/Makefile:431]: I will try to update only the FSP to the latest version, now v4.4.0.   I think I can achieve this by `git submodule update --remote lib/fsp`.   Is this operation appropriate?

Comment (iabdalkader) [reply ports/renesas-ra/Makefile:431]: This is untested, but it should be something like:

```bash
cd lib/fsp
git fetch origin
git checkout v4.4.0
cd ../
git add fsp
git commit -m "lib/fsp: update fsp to version 4.4.0."
```

Any link to documentation on how to update the fsp pack in e2 studio ?

Comment (TakeoTakahashi2020) [reply ports/renesas-ra/Makefile:431]: Thanks @iabdalkader san.  After updating, I tried to build but many compile errors happen.  Current version is very old, so it looks need to update ra_gen/* and others.  It might be not easy than I guess unfortunattely. 

>Any link to documentation on how to update the fsp pack in e2 studio ?

I hope the following FAQ link helps: 
https://en-support.renesas.com/knowledgeBase/20612472?_ga=2.71050256.1205374166.1682471818-1570124239.1633348365&_gac=1.183302740.1679888466.Cj0KCQajw2v-gBhC1ARIsAOQdKY3wL4ZrIc__3Btgo0mxi8wc2_1ci1Qqmyl6HhqpM1Rh_mflR35li1IaAn2vEALw_wcB&_gl=1*l9lvmi*_ga*MTU3MDEyNDIzOS4xNjMzMzQ4MzY1*_ga_D1706WVDQV*MTY4MjU5NDkwOC42ODEuMS4xNjgyNTk2MDYxLjAuMC4w

Comment (iabdalkader) [reply ports/renesas-ra/Makefile:431]: Thank you that's exactly what I was looking for.

I agree it is a lot of work to update all of the projects, perhaps at some point in the future dependency on e2 studio generated files can be removed, but I'm okay with the current solution so it's up to @dpgeorge now.

Comment (dpgeorge) [reply ports/renesas-ra/Makefile:431]: > but I'm okay with the current solution

Sorry I don't quite follow.  What is the "current solution" that you refer to?

Comment (iabdalkader) [reply ports/renesas-ra/Makefile:431]: > Sorry I don't quite follow. What is the "current solution" that you refer to?

The current solution, which is in this PR, is to switch the `lib/fsp` version on the fly.

Comment (iabdalkader) [reply ports/renesas-ra/Makefile:431]: I agree not having a dependency on e2 studio generated files at all would be great, but I don't know enough to say for sure if it's not needed, to me it seems that only the vector tables and the headers that enable/config different stacks are really needed, not sure if the other generated data structures are used.

**Comment by dpgeorge** [comment]
@mbedNoobNinja can you please rebase this on the latest master, now that the FSP is updated to 4.4.0?

**Comment by mbedNoobNinja** [comment]
Hi @dpgeorge I the PR is rebased.

**Comment by TakeoTakahashi2020** [comment]
Thanks @mbedNoobNinja.  I have confirmed that buid for VK-RA6M5, all EKs,  CLICKER is fine, and I2C, SPI and so on seems to be working well.

**Comment by iabdalkader** [comment]
Unrelated issue, I'm trying to use UART with baudrate > 115200 (for ESP bluetooth firmware which uses 921600) and noticed that higher baudrates are not supported, so I copied `R_SCI_UART_BaudCalculate` from FSP and it seems to be working fine (tried 115200, and 921600). 

@ everyone What would be the best to go about fixing this ? `R_SCI_UART_BaudCalculate` seems to be pretty much standalone, can we reuse it somehow ? Link the file that provides or just add a copy in renesas-ra/ra/ra_sci.c` ?  CC @dpgeorge

**Comment by TakeoTakahashi2020** [comment]
@iabdalkader -san, I tried to link the file, however, duplicated symbol error happens for interrupt function.  Then I tried SCI_UART_CFG_TX_ENABLE (0)/SCI_UART_CFG_RX_ENABLE (0) configuration but  unused-variable error happens.  Seems to be needed more consideration and drastic changes to link the file. 
If you had obtained the result of R_SCI_UART_BaudCaluclate for bluetooth, we might be able to try to use only the result in ra/ra_sci.c as the first step to keep ROM/RAM size small.  However, this is one of idea.

**Comment by iabdalkader** [comment]
> If you had obtained the result of R_SCI_UART_BaudCaluclate for bluetooth

I did but I think they are board/clock specific and probably shouldn't be hard-coded. Can we just add a copy of that function to `ra/ra_sci.c` ? I already have an open PR can just add it there, will format it and remove any unused code.

**Comment by mbedNoobNinja** [comment]
Added SDHI support in `machine.SDCard` and frozen SDCard module is removed. Regarding the baudrate I think **R_SCI_UART_BoudCalculate** could be copied in `ra_sci.c` in some form in the existing function **ra_sci_set_baud()** and then the existing implementation of the uPy method **UART.init(baudrate)** should work properly. But that of course conserns all RA bords and probably the fix shuld be in a separate PR.

**Comment by iabdalkader** [comment]
I already copied it, with some minor formatting cleanups and will add it to my open PR later. Note I tested changing the baudrate and everything seems to be working fine.

**Comment by TakeoTakahashi2020** [comment]
> Can we just add a copy of that function to `ra/ra_sci.c` ? 

Thank you for waiting.  I was still trying linking the r_sci_uart.c of FSP and calling R_SCI_UART_BaudCalculate() because I am concering about FSP license.

Some modifications for `Makefile`, `ra/ra_sci.c `and `boards/*/ra_cfg/fsp_cfg/r_sci_uart_cfg.h` were needed, but I could support baudrate change and UART.init(baudrate).  This increase ROM size of 384 bytes.  Now I am cheking and it is planned to open a new PR separately tomorrow (Japan time), so I would be happy if you all could review it.

**Comment by iabdalkader** [comment]
Okay will review and test it, as long as we can change the baudrate it's fine.

**Comment by dpgeorge** [comment]
This PR is looking a lot better now, but I still want to further simplify and reduce the files in each board directory.

See #11590 for the first PR to do this, which is merged.

And see #11846 for the second set of simplifications.  I think that will help simplify this PR as well.

**Comment by dpgeorge** [comment]
I have rebased this on latest master and added some commits which:
- consolidate the fsp_cfg files, as per previous commits
- clean up some minor formatting issues
- rename DAC `write_mV` and `read_mV` to `write_mv` and `read_mv` (we don't use upper case in method/function names)
- remove escape codes from error messages, and simplify some messages
- change ValueError to OSError for SD card errors

**Comment by iabdalkader** [comment]
Excellent, thank you!

**Comment by dpgeorge** [comment]
Rebased again on latest master, to pick up the improvements to the UART.

From my point of view, this looks OK to merge now (commits will be squashed before merging).

**Comment by dpgeorge** [comment]
> I'm probably nitpicking but it would look better split into 2 commits

Yes I agree.  I'll try to do that.

**Comment by dpgeorge** [comment]
Thanks @mbedNoobNinja for your work on this.

**Comment by TakeoTakahashi2020** [comment]
Great work @mbedNoobNinja.  I am happy to add new RA board.

**Pull request merged**
